libtenzir/src/secret_resolution.cpp

```diff
@@ -0,0 +1,162 @@
+    reinterpret_cast<const uint8_t*>(value_.data()), value_.size());
+  if (not valid_utf8) {
+    return std::nullopt;
+  }
+  return std::string_view{
+    reinterpret_cast<const char*>(value_.data()),
+    value_.size(),
+  };
+}
+
+auto resolved_secret_value::utf8_view(std::string_view name, location loc,
+                                      diagnostic_handler& dh) const
+  -> failure_or<std::string_view> {
+  auto r = utf8_view();
+  if (not r) {
+    diagnostic::error("expected secret `{}` to be a UTF-8 string", name)
+      .primary(loc)
+      .emit(dh);
+    return failure::promise();
+  }
+  return *r;
+}
+
+namespace detail {
+
+auto secret_resolved_setter_callback(resolved_secret_value& out) {
+  return [&out](resolved_secret_value v) -> failure_or<void> {
+    out = std::move(v);
+    return {};
+  };
+}
+
+auto secret_string_setter_callback(std::string name, tenzir::location loc,
+                                   std::string& out, diagnostic_handler& dh)
+  -> secret_request_callback {
+  return [name, loc, &out, &dh](resolved_secret_value v) -> failure_or<void> {
+    TRY(auto str, v.utf8_view(name, loc, dh));
+    out = std::string{str};
+    return {};
+  };
+}
+
+auto secret_string_setter_callback(std::string name, tenzir::location loc,
+                                   located<std::string>& out,
+                                   diagnostic_handler& dh)
+  -> secret_request_callback {
+  return [name, loc, &out, &dh](resolved_secret_value v) -> failure_or<void> {
+    TRY(auto str, v.utf8_view(name, loc, dh));
+    out = located{std::string{str}, loc};
+    return {};
+  };
+}
+
+} // namespace detail
+
+/// A secret request that will invoke `callback` on successful resolution
+secret_request::secret_request(tenzir::secret secret, tenzir::location loc,
+                               resolved_secret_value& out)
+  : secret{std::move(secret)},
+    location{loc},
+    callback{detail::secret_resolved_setter_callback(out)} {
+}
+
+secret_request::secret_request(const located<tenzir::secret>& secret,
+                               resolved_secret_value& out)
+  : secret{std::move(secret.inner)},
+    location{secret.source},
+    callback{detail::secret_resolved_setter_callback(out)} {
+}
+
+auto secret_censor::censor(std::string text) const -> std::string {
+  constexpr static auto replacement = std::string_view{"***"};
+  auto single_star = false;
+  auto double_star = false;
+  for (const auto& s : secrets) {
+    const auto v
+      = std::string_view{reinterpret_cast<const char*>(s.data()), s.size()};
+    if (v.empty()) {
+      continue;
+    }
+    if (v == "*") {
+      single_star = true;
+      continue;
+    }
+    if (v == "**") {
+      double_star = true;
+      continue;
+    }
+    if (v == "***") {
+      continue;
+    }
+    for (auto p = text.find(v); p < text.size(); p = text.find(v, p)) {
+      text.replace(p, v.size(), replacement);
+      p += replacement.size();
+    }
```

Comment:
One more thing that just randomly thought about: If you replace strings like this and one of the secrets was a blob, then you may end up partially replacing codepoints and therefore get invalid UTF-8 here… It's an absolute definition of an edge case and maybe we just ignore it, but just to have it mentioned…